libpromises/pipes_unix.c

```diff
@@ -448,27 +486,25 @@ FILE *cf_popensetuid(const char *command, const char *type, uid_t uid, gid_t gid
             }
         }
 
-        CloseChildrenFD();
-
-        argv = ArgSplitCommand(command);
+        ChildrenFDUnsafeClose();
 
         if (chrootv && (strlen(chrootv) != 0))
         {
             if (chroot(chrootv) == -1)
             {
                 Log(LOG_LEVEL_ERR, "Couldn't chroot to '%s'. (chroot: %s)", chrootv, GetErrorStr());
-                ArgFree(argv);
-                return NULL;
+                _exit(EXIT_FAILURE);
             }
         }
 
         if (chdirv && (strlen(chdirv) != 0))
         {
+            /* TODO: safe_chdir is probably not signal-handler safe, so it's
+             * not allowed in child. */
             if (safe_chdir(chdirv) == -1)
```

Comment:
It's completely correct, and it's not "probably". It *isn't* signal-handler safe. :-)

Reply:
Do you think plain `chdir()` would be appropriate in this context?

Reply:
Uhm, hold on a minute. Actually no. It's true that the function `safe_chdir` is not itself signal safe, but it still is due to the way we use it. See [this comment](https://github.com/cfengine/core/blob/master/libcompat/generic_at.c#L39-L53). Due to cf-agent being single threaded, calling `safe_chdir` should be safe, because it can only be this thread that holds the lock (and at this point in the code it certainly won't). And this is probably a place we *should* call `safe_chdir`, because the execution argument might come from a promise. So I take back my earlier comment.

Reply:
You are right, this is the exact path that should be guarded with `safe_open()`. But many functions in this file are being used by serverd and execd, and these programs are multithreaded. I'll audit if the specific functions that call `safe_*()` are only called by the agent. 

`safe_open()` should most likely assert that the program is single threaded, if such a thing is possible.